asserts/account_key_test.go

```diff
@@ -419,3 +421,61 @@ func (aks *accountKeySuite) TestPrerequisites(c *C) {
 		PrimaryKey: []string{"acc-id1"},
 	})
 }
+
+func (aks *accountKeySuite) TestKeyProofHappy(c *C) {
+	kproof, err := asserts.SignWithoutAuthority(asserts.KeyProofType,
+		map[string]interface{}{
+			"nonce": "NONCE",
+			"key":   aks.pubKeyEncoded,
+		}, nil, aks.privKey)
+	c.Assert(err, IsNil)
+
+	// roundtrip
+	a, err := asserts.Decode(asserts.Encode(kproof))
+	c.Assert(err, IsNil)
+
+	kproof2, ok := a.(*asserts.KeyProof)
+	c.Assert(ok, Equals, true)
+
+	// standalone signature check
+	err = asserts.SignatureCheck(kproof2, aks.privKey.PublicKey())
+	c.Check(err, IsNil)
+
+	c.Check(kproof2.Nonce(), Equals, "NONCE")
+}
+
+func (aks *accountKeySuite) TestKeyProofDecodeInvalid(c *C) {
+	encoded := "type: key-proof\n" +
+		"nonce: NONCE\n" +
+		"key:\n    KEY\n" +
+		"body-length: 0\n" +
+		"sign-key-sha3-384: " + aks.privKey.PublicKey().ID() + "\n\n" +
+		"AXNpZw=="
+
+	invalidTests := []struct{ original, invalid, expectedErr string }{
+		{"nonce: NONCE\n", "", `"nonce" header is mandatory`},
+		{"nonce: NONCE\n", "nonce: \n", `"nonce" header should not be empty`},
+		{"key:\n    KEY\n", "", `"key" header is mandatory`},
+		{"key:\n    KEY\n", "key: \n", `"key" header should not be empty`},
+		{"key:\n    KEY\n", "key: $$$\n", `cannot decode public key: .*`},
+	}
+
+	for _, test := range invalidTests {
+		invalid := strings.Replace(encoded, test.original, test.invalid, 1)
+		invalid = strings.Replace(invalid, "KEY", strings.Replace(aks.pubKeyEncoded, "\n", "\n    ", -1), 1)
+		_, err := asserts.Decode([]byte(invalid))
+		c.Check(err, ErrorMatches, keyProofErrPrefix+test.expectedErr)
+	}
+}
+
+func (aks *accountKeySuite) TestKeyProofDecodeKeyIDMismatch(c *C) {
+	invalid := "type: key-proof\n" +
```

Comment:
It clarifies, and I appreciate the insight into all these details, but the explanation clearly indicates a misuse of assertions. It would be nice to do one of two things: either make the assertion an actual assertion, which provides meaningful data, or stop using an assertion and just sign a nonce. My preference is actually to hold it as an assertion, and make it a relevant assertion that has historical value instead of being a wire-protocol gimmick.

Let me provide better rationale:

> Of course the key-proof assertion itself only proves that whoever signed the assertion has access to the private key. That's its purpose, and that's why this assertion is only intended for use on the wire and not to be stored long-term.

Right, exactly. That's a fine thing to do on a wire-protocol exchange which has a very tight context (client is authenticated, timing is extremely limited, etc), but as an assertion its completely uninteresting. We cannot even use it to justify to the private key owner that there was indeed the intention of publishing that key, or which account it was supposed to be bound to. 

This is the beauty of assertions. They enable us to backtrack and justify decisions taken in the store based on data which we did not author, and we can use that data to prove to whoever asks (devices, third-parties, etc) that the decision we took was the right one based on facts.

If we craft this assertion a bit more carefully, we should be able to store it as proof that there was a real intention of someone in possession of that private key to publish its public side in the store, at that time, under that specific account id.

We need to rename the assertion for that. It must be more indicative of the intended outcome, rather than being so passive. We can bikeshed on a name, but we'll save time by doing that interactively online.

We don't even need a nonce, I believe. The key ID itself is the nonce. We must never take a second one, and the first one must hold good acceptable data.